File: _tools/check-doc-links.py
"""Check local destinations and fragments in the built documentation."""
from html.parser import HTMLParser
from pathlib import Path
from urllib.parse import unquote, urlsplit


class Links(HTMLParser):
    """Collect element IDs and anchor destinations from one HTML page."""

    def __init__(self, source):
        """Parse page text into ID and link collections."""
        super().__init__()
        self.ids = set()
        self.links = []
        self.feed(source)

    def handle_starttag(self, tag, attrs):
        """Record IDs on any element and href values on anchors."""
        attrs = dict(attrs)
        if "id" in attrs:
            self.ids.add(attrs["id"])
        if tag == "a" and "href" in attrs:
            self.links.append(attrs["href"])


def local_target(root, source, url_path):
    """Resolve a decoded URL path and directory index relative to its page or site root."""
    destination = unquote(url_path)
    if not destination:
        target = source
    elif destination.startswith("/"):
        target = root / destination.lstrip("/")
    else:
        target = source.parent / destination
    if target.is_dir():
        target = target / "index.html"
    return target.resolve()


def check_link(root, source, href, pages):
    """Return the first local-link diagnostic, or None for a valid or external link."""
    link = urlsplit(href)
    if link.scheme or link.netloc:
        return None
    target = local_target(root, source, link.path)
    location = source.relative_to(root)
    if not target.is_relative_to(root):
        return f"{location}: outside output: {href}"
    if not target.exists():
        return f"{location}: missing destination: {href}"
    if link.fragment and target in pages and unquote(link.fragment) not in pages[target].ids:
        return f"{location}: missing anchor: {href}"
    return None


def check_site(root):
    """Scan built pages and fail with collected link diagnostics, or report the page count."""
    pages = {path: Links(path.read_text()) for path in root.rglob("*.html")}
    if not pages:
        raise SystemExit("No built pages found. Run bun run docs:build first.")
    errors = []
    for source, page in pages.items():
        for href in page.links:
            error = check_link(root, source, href, pages)
            if error is not None:
                errors.append(error)
    if errors:
        raise SystemExit("\n".join(errors))
    print(f"Checked local links and fragments in {len(pages)} pages.")


if __name__ == "__main__":
    check_site(Path(__file__).resolve().parents[1] / "docs" / "dist")

File: docs/_tools/check-doc-links.ts
/** @fileoverview Checks local destinations and fragments in built documentation using parsed HTML. */

import {
  existsSync,
  readFileSync,
  readdirSync,
  realpathSync,
  statSync,
} from 'node:fs';
import {
  basename,
  dirname,
  isAbsolute,
  join,
  relative,
  resolve,
  sep,
} from 'node:path';
import { unescape as decodePercentEscapes } from 'node:querystring';
import { parse } from 'parse5';
import type { DefaultTreeAdapterTypes } from 'parse5';

/** IDs and anchor destinations found in a built page, after HTML entity decoding. */
type PageLinks = {
  readonly ids: ReadonlySet<string>;
  readonly links: ReadonlyArray<string>;
};
/** A local URL's path and fragment, with query parameters omitted. */
type LocalReference = { readonly path: string; readonly fragment: string };

/** Collect IDs and anchor href values; comments and script text are not treated as markup. */
function pageLinks(html: string): PageLinks {
  const ids = new Set<string>();
  const links: Array<string> = [];
  /** Visit parsed elements and their children in document order. */
  function visit(node: DefaultTreeAdapterTypes.Node): void {
    if ('tagName' in node) {
      const id = node.attrs.find((attribute) => attribute.name === 'id');
      if (id !== undefined) ids.add(id.value);
      if (node.tagName === 'a') {
        const href = node.attrs.find((attribute) => attribute.name === 'href');
        if (href !== undefined) links.push(href.value);
      }
    }
    if ('content' in node) visit(node.content);
    if ('childNodes' in node) for (const child of node.childNodes) visit(child);
  }
  visit(parse(html));
  return { ids, links };
}

/** Discover HTML files without following symlinked directories outside the built site. */
function htmlFiles(root: string): ReadonlyArray<string> {
  const paths: Array<string> = [];
  /** Walk physical subdirectories only; symlink destinations are checked when referenced by a link. */
  function visit(directory: string): void {
    for (const entry of readdirSync(directory, { withFileTypes: true })) {
      if (entry.isSymbolicLink()) continue;
      const path = join(directory, entry.name);
      if (entry.isDirectory()) visit(path);
      else if (entry.isFile() && entry.name.endsWith('.html')) paths.push(path);
    }
  }
  if (existsSync(root)) visit(root);
  return paths.sort();
}

/** Load built HTML pages in path order, returning no pages when the build directory is absent. */
function builtPages(root: string): ReadonlyMap<string, PageLinks> {
  return new Map(
    htmlFiles(root).map((path) => [
      path,
      pageLinks(readFileSync(path, 'utf8')),
    ]),
  );
}

/** Split local references without URL normalization hiding traversal outside the output root. */
function localReference(rawHref: string): LocalReference | undefined {
  const href = rawHref.replace(/^[\x00-\x20]+/u, '').replace(/[\t\r\n]/gu, '');
  if (/^[a-z][a-z0-9+.-]*:/iu.test(href) || href.startsWith('//'))
    return undefined;
  const hash = href.indexOf('#');
  const beforeHash = hash < 0 ? href : href.slice(0, hash);
  const query = beforeHash.indexOf('?');
  return {
    path: query < 0 ? beforeHash : beforeHash.slice(0, query),
    fragment: hash < 0 ? '' : href.slice(hash + 1),
  };
}

/** Resolve existing symlinks while retaining the unresolved suffix of a missing destination. */
function resolvedPath(path: string): string {
  if (existsSync(path)) return realpathSync(path);
  const parent = dirname(path);
  if (parent === path) return path;
  return join(resolvedPath(parent), basename(path));
}

/** Resolve a local URL from its page or the site root, including directory index files. */
function localTarget(root: string, source: string, pathname: string): string {
  // Unlike form decoding, URL paths retain literal plus signs; malformed percent escapes remain literal.
  const destination = decodePercentEscapes(pathname);
  let target: string;
  if (destination === '') target = source;
  else if (destination.startsWith('/'))
    target = resolve(root, `.${destination}`);
  else target = resolve(dirname(source), destination);
  const canonical = resolvedPath(target);
  return existsSync(canonical) && statSync(canonical).isDirectory()
    ? resolvedPath(join(canonical, 'index.html'))
    : canonical;
}

/** Report the first failure for a local link, or return nothing for a valid or external link. */
function checkLink(
  root: string,
  source: string,
  href: string,
  pages: ReadonlyMap<string, PageLinks>,
): string | undefined {
  const link = localReference(href);
  if (link === undefined) return undefined;
  const target = localTarget(root, source, link.path);
  const relativeTarget = relative(root, target);
  const location = relative(root, source);
  if (
    relativeTarget === '..' ||
    relativeTarget.startsWith(`..${sep}`) ||
    isAbsolute(relativeTarget)
  )
    return `${location}: outside output: ${href}`;
  if (!existsSync(target)) return `${location}: missing destination: ${href}`;
  const targetPage = pages.get(target);
  if (
    link.fragment &&
    targetPage !== undefined &&
    !targetPage.ids.has(decodePercentEscapes(link.fragment))
  )
    return `${location}: missing anchor: ${href}`;
  return undefined;
}

/** Check every local link and return a success summary; throw with all link diagnostics when any fail. */
function checkSite(root: string): string {
  const pages = builtPages(root);
  if (pages.size === 0)
    throw new Error('No built pages found. Run bun run docs:build first.');
  const errors: Array<string> = [];
  for (const [source, page] of pages) {
    for (const href of page.links) {
      const error = checkLink(root, source, href, pages);
      if (error !== undefined) errors.push(error);
    }
  }
  if (errors.length) throw new Error(errors.join('\n'));
  return `Checked local links and fragments in ${pages.size} pages.`;
}

if (import.meta.main) {
  try {
    const root = resolvedPath(
      resolve(process.argv[2] ?? join(import.meta.dir, '../dist')),
    );
    console.log(checkSite(root));
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    process.exitCode = 1;
  }
}
